### ISSUE TITLE
Disable Forms Authentication Redirect when returning HTTP 401

### ISSUE THREAD

**Issue by mschmidt98** [dotnet/aspnetcore#27805]
Hi,

**TL;DR:** Is there an ASPNET.Core alternative to [HTTPResponse.SuppressFormsAuthenticationRedirect](https://docs.microsoft.com/en-us/dotnet/api/system.web.httpresponse.suppressformsauthenticationredirect), so that a Website that has Forms Authentication enabled and consumes the API does not redirect to the login, but returns the error the API provides?

**Long Version with explanation**:
We have an ASP.NET Core Webserver on Version 2.2 on .NET Framework. Due to external dependancies we can not migrate to .NET Core and Version 3 at the moment. This Webserver combines multiple APIs and is supposed to be hosted inside the private part of a network for security reasons.

Now, some of the APIs need to be exposed externally, so we provide an IIS website as some kinde of proxy, that is supposed to be hosted inside the DMZ. For administraiton purposes, this websites contains a login area and has forms authentication enabled.

We added an API, that is part of the internally hosted server, and might return a 401 code when the authentication fails. Via URL rewrite, this is made public via mentioned IIS site.

When calling the endpoint via the internal url, the result is just as expected:
![image](https://user-images.githubusercontent.com/12734928/99063834-69cff500-25a5-11eb-9201-0311e2db0b73.png)

Now, when calling the endpoint externally, the response is changed inside IIS to the following, so that instead of the 401 with our custom error message, the proxy site returns status 200 with the html code of the login page as content. For a public programming api, this removes important information regarding the error.
![image](https://user-images.githubusercontent.com/12734928/99064134-cfbc7c80-25a5-11eb-95c4-d1abe554dd7a.png)

After disabling Forms Authentication in the IIS proxy site, the result gets passed through as expected again:
![image](https://user-images.githubusercontent.com/12734928/99064364-2924ab80-25a6-11eb-9da9-3222db2a355b.png)

From my understanding, to keep the IIS from interpreting the response itself and changing it, the [HTTPResponse.SuppressFormsAuthenticationRedirect](https://docs.microsoft.com/en-us/dotnet/api/system.web.httpresponse.suppressformsauthenticationredirect) flag was added in .NET Framework 4.5, but I was not able to find out, how to achieve this behaviour with ASP.NET Core, where we just migrated to some months ago, to get rid of legacy MVC dependancies.

Is this behaviour also achievable with ASP.NET Core 2.2, maybe using some sort of middleware? Or would we have to either change the status code or somehow get rid of forms authentication?

Thank you and best regards,
Markus

**Comment by blowdart** [comment]
I'm a little confused about your layout here.

You say you have an IIS server in a DMZ that acts as a proxy. Is it actually a proxy, where it terminates and forwards requests to an internal server, or is it running the api service directly?

Also what is calling this api at the DMZ? A website, an app?

**Comment by mschmidt98** [comment]
Sorry if my explanation was unclear.
Proxy might be to wide of a term. The requests regarding our API are just forwarded using URL rewrite, simple as that. But because this IIS site also contains some configuration sites with UI, the forms authentication is acitvated and needed.

The actual logic is only at the internal service, where it is also used for some internal web apps.
Externally, the code might be called by any system, since it is an API meant for public use. Might be HR or applicant management software, that might not run in the same network as our services - that is why we want the error codes to be as specific as possible.

**Comment by Tratcher** [comment]
No, there's no support for this today, you're managing the authentication completely at the IIS layer.

If the authentication was being managed at the application layer then you'd get some control over this. Unfortunately there's no integrated support for Active Directory users via forms auth in Asp.Net Core. You'd need to shift to something like WsFed or OpenIdConnect.

**Comment by blowdart** [comment]
Closing as by design/external.
